Escaping whitespace in arguments

seinmon opened the pull request:
<!-- Provide a descriptive summary of the changes in the title above -->

#### ISSUE TYPE
<!-- Pick relevant types and delete the rest -->
- Bug fix

#### RUNTIME ENVIRONMENT
<!-- Details of your runtime environment -->
<!-- Retrieve Python/ranger version and locale with `ranger -\-version` -->
- Operating system and version: Arch Linux
- Terminal emulator and version: urxvt v9.26, (+ tmux 3.2a)
- Python version: 3.10.1
- Ranger version/commit: ranger-master v1.9.3
- Locale: None.None

#### CHECKLIST
<!-- All [REQUIRED] requisites need to be fulfilled -->
<!-- Replace [ ] with [X] when fulfilled -->
- [x] The `CONTRIBUTING` document has been read **[REQUIRED]**
- [x] All changes follow the code style **[REQUIRED]**
- [x] All new and existing tests pass **[REQUIRED]**
- [ ] Changes require config files to be updated
    - [ ] Config files have been updated
- [ ] Changes require documentation to be updated
    - [ ] Documentation has been updated
- [ ] Changes require tests to be updated
    - [ ] Tests have been updated

#### DESCRIPTION
<!-- Describe the changes in detail -->
This patch adds the possibility of escaping of space character as "\ " to arguments.

#### MOTIVATION AND CONTEXT
<!-- Why are these changes required? -->
<!-- What problems do these changes solve? -->
<!-- Link to relevant issues -->
fixes #2470 
This is useful for scenarios where the a directory name containing whitespace is passed as an argument.
 
#### TESTING
<!-- What tests have been run? -->
<!-- How does the changes affect other areas of the codebase? -->
The code is tested with `make test_py` and the output shows `rated 9.85/10`.

#### IMAGES / VIDEOS<!-- Only if relevant -->
<!-- Link or embed images and videos of screenshots, sketches etc. -->

toonn commented:
I just pushed updates that should fix the Pylint errors. I'm not very confident this works as-is. Escaping is really hard to handle well. What we really need is to be able to quote the argument. What if the destination literally contains `\ ` for example?

seinmon commented:
I see! This already works with destinations that have "\" in their name, but it doesn't work if the name has "\ " in it.  (Apparently `mv` also doesn't seem to work with these kinds of names). I can also try wrapping the destination in quotation marks.

Since the codebase is rather large, are there other places that need to be changed for this patch to work as expected?

seinmon commented:
@toonn I tried to quote the arguments, but it becomes unnecessarily complicated. I can try to finish quoting the argument if you want, but I have to ask this first:

What if the name actually contains `"`? Wrap the argument in `'`? What if it contains `'`? Wrap it in `"`? What if it contains both? Escape them? 

I guess the chance of a name containing quotation marks are a lot higher than a name containing `\ ` (slash + space) as `\ ` is how we escape spaces in other commands like `mv`, which also doesn't handle names containing `\ ` (and this behavior probably makes escaping more obvious/familiar to the users).

Please let me know what you think.

tobiasjakobi commented:
How about using shlex.split()? This should handle quotes and escaped whitespaces.

seinmon commented:
@tobiasjakobi thank you for the great suggestion. I tried `shlex.split()`, but `ranger` tries to split commands while the user is typing. As a result, `shlex.split()` gets called with no closing quotation marks\escaped characters and crashes. (I haven't had enough time to look for a workaround.)

tobiasjakobi commented:
@seinmon I'm not sure what you mean by crash? I tried some input where closing quotations were missing, and I don't see any crashes of the interpreter.

seinmon commented:
@tobiasjakobi When I give an input like `"something` (missing closing quotation) or `some\` (missing escaped character) then the script ends with a `ValueError` exception. Isn't this the same for you?

tobiasjakobi commented:
Of course, but that has nothing to do with any kind of "crashing". It's just a regular Python exception (which you need to handle).

seinmon commented:
@tobiasjakobi Now I feel stupid! I forgot that I need to handle the exception! 🤣 It works perfectly! Thanks. I will commit changes and push them to my forked repository.

seinmon commented:
@toonn Can you please pull the changes again? (`Pylint` tests fail for files that I haven't modified.)

seinmon commented:
Up @toonn

toonn reviewed (commented):
I'm a bit fearful this is going to wreak havoc with commands like `shell` which need to pass things to the shell literally because the shell will do something that `shlex.split` tries to mimic.
